### PR TITLE
[Dependent] Simple Review Pipeline (Client)

### LINKED ISSUE #2082: Find a better place to job status switcher
### My actions before raising this issue
- [X] Read/searched [the docs](https://github.com/opencv/cvat/tree/master#documentation)
- [x] Searched [past issues](/issues)

I don't have an option to change status of job.
Screen Shot 2020-08-25 at 8 52 48 AM
Additionally the enumeration type for the status of the job in enums.js is called TaskStatus. I think the name of this should be changed since it is referring to a task not a job.

### Expected Behaviour

The menu on the annotation page gives option to change status.
### Steps to Reproduce (for bugs)

1.
1.
1.
1.

### Your Environment

- Git hash commit (`git log -1`): 1fdf10032a072da9f3ac19bfd623accc793c88cb
- Docker version `docker version` (e.g. Docker 17.0.05):
- Are you using Docker Swarm or Kubernetes?
- Operating System and version (e.g. Linux, Windows, MacOS): MacOS
- Code example or link to GitHub repo or gist to reproduce problem:
- Other diagnostic information / logs:
    <details>
    <summary>Logs from `cvat` container</summary>
    </details>

### Next steps
You may [join our Gitter](https://gitter.im/opencv-cvat/public) channel for community support.

### PR DESCRIPTION
**Pull request by bsekachev**
### Motivation and context
Depends on #2338 
Resolved #2082
Related #1100

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2020 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

**Review by ActiveChooN — COMMENTED** [review]
Great job! Have some thoughts.

Need to add "ago" here
изображение

Guess it's better to print here only first comment, not last
изображение

If i send annotation it will redirect me to the task page, but if i submit a review it will not. Is it normal behavior?

If task have all jobs in validation status, task status would not updated in UI. Guess it will be better to have complex progress bar with blue line for annotated jobs and green for successfully reviewed jobs. Also if i submit annotation with accepted status and go to the tasks page and job status in job list block was not updated. I noticed same behavior with reviewer value. If i choose "review next" option reviewer doesn't change in job list block even if i update page.
изображение
изображение

Maybe it could be more convenient to set up review buttons with next order "Accept", "Review next", "Decline", but I can be wrong.

Also should be current user available for assigning for next review?

And i'm not pretty sure about blinking animation for selected roi. Maybe it's better to use white, because it's not used in label color selection algorithm.

File: cvat-ui/src/components/annotation-page/review/submit-review-modal.tsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+                                <td>{reviewedFrames}</td>
+                            </tr>
+                            <tr>
+                                <td>
+                                    <Text strong>Reviewed objects: </Text>
+                                </td>
+                                <td>{reviewedStates}</td>
+                            </tr>
+                        </tbody>
+                    </table>
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/components/annotation-page/review/submit-review-modal.tsx:115]: Great idea, now it is used here.

File: cvat-ui/src/components/annotation-page/review/issues-aggregator.tsx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+    useEffect(() => {
+        const listener = (): void => scaleHandler(canvasInstance);
+
+        canvasInstance.html().addEventListener('canvas.zoom', listener);
+        canvasInstance.html().addEventListener('canvas.fit', listener);
+
+        return () => {
+            canvasInstance.html().addEventListener('canvas.zoom', listener);
+            canvasInstance.html().addEventListener('canvas.fit', listener);
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-ui/src/components/annotation-page/review/issues-aggregator.tsx:51]: Yes, there was the issue. Fixed

File: cvat-core/src/issue.js [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+                 * @memberof module:API.cvat.classes.Issue
+                 * @readonly
+                 * @instance
+                 */
+                resolver: {
+                    get: () => data.resolver,
+                },
+                /**
+                 * @name removed
+                 * @type {module:API.cvat.classes.Comment}
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-core/src/issue.js:153]: Fixed

File: cvat-core/src/comment.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                 * @memberof module:API.cvat.classes.Comment
+                 * @readonly
+                 * @instance
+                 */
+                author: {
+                    get: () => data.author,
+                },
+                /**
+                 * @name removed
+                 * @type {module:API.cvat.classes.Comment}
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-core/src/comment.js:104]: Yes, fixed

File: cvat-core/src/comment.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                 * @readonly
+                 * @instance
+                 */
+                updatedDate: {
+                    get: () => data.updated_date,
+                },
+                /**
+                 * Instance of a user who has created the comment
+                 * @name author
+                 * @type {module:API.cvat.classes.Comment}
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-core/src/comment.js:94]: Yes, fixed

File: cvat-canvas/src/typescript/regionSelector.ts [hunk 6]
```diff
@@ -0,0 +1,190 @@
+// Copyright (C) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
+
+// Copyright (C) 2019-2020 Intel Corporation
+//
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (bsekachev) [reply cvat-canvas/src/typescript/regionSelector.ts:7]: Removed extra header

**Comment by bsekachev** [comment]
> Need to add "ago" here

Agree. Has been added.

> Guess it's better to print here only first comment, not last 

Not sure here. IMHO the message might be considered like an indicator if there are any updates in dialog or not for a user

> Maybe it could be more convenient to set up review buttons with next order "Accept", "Review next", "Decline", but I can be wrong.

Yes. Done.

> Also should be current user available for assigning for next review?

In the most of cases it doesn't make sense, but I do not see reasons why we should strict it implicitly. 

> And i'm not pretty sure about blinking animation for selected roi. Maybe it's better to use white, because it's not used in label color selection algorithm.

Firstly it had been implemented exactly so, but on sometimes (especially when issue region is a point) it rather hard to see the region on the canvas.

**Comment by bsekachev** [comment]
> If task have all jobs in validation status, task status would not updated in UI.

In UI it has a bit different meaning. It is pending if no one of jobs completed, if at least on of jobs completed it is in progress, if all the jobs are completed, the task is completed.

**Comment by bsekachev** [comment]
Other problems should be fixed now

**Comment by bsekachev** [comment]
@ActiveChooN 

Could you please look at the patch again?

**Comment by ActiveChooN** [comment]
> @ActiveChooN 
> 
> Could you please look at the patch again? 

Sure

**Comment by nmanovic** [comment]
Try to build it locally and got the following error: 

```
Step 20/29 : RUN npm ci
 ---> Running in d3865027b8c7
npm ERR! code EEXIST
npm ERR! syscall mkdir
npm ERR! path /tmp/cvat-ui/node_modules/cvat-core/node_modules/cvat-data
npm ERR! errno -17
npm ERR! EEXIST: file already exists, mkdir '/tmp/cvat-ui/node_modules/cvat-core/node_modules/cvat-data'
npm ERR! File exists: /tmp/cvat-ui/node_modules/cvat-core/node_modules/cvat-data
npm ERR! Remove the existing file and try again, or run npm
npm ERR! with --force to overwrite files recklessly.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-11-14T07_10_17_653Z-debug.log
ERROR: Service 'cvat_ui' failed to build : The command '/bin/sh -c npm ci' returned a non-zero code: 239
```

**Comment by ActiveChooN** [comment]
Also related the to server PR, but could you please fix issues with server tests?

**Comment by ActiveChooN** [comment]
@bsekachev 

изображение

I cannot close `Issue #1` with some isuues openned, maybe it's better to have only one opened issue window.

**Comment by bsekachev** [comment]
> I cannot close Issue #1 with some isuues openned

Let's try up a dialog window on mouseover in DOM

**Comment by ActiveChooN** [comment]
Some other thoughts about issue popup:
- It can be realy useful to send comment with `Enter` button
- Box-shadows can help to identify issue popup more easily

**Comment by bsekachev** [comment]
@ActiveChooN 

They are nice ideas, added.

**Comment by ActiveChooN** [comment]
Can we use [default](https://github.com/ant-design/ant-design/blob/master/components/style/themes/default.less) shadows from antd?

**Comment by bsekachev** [comment]
@ActiveChooN 

We can specify one of the same styles in our code, but we cannot import less file into sass.

**Comment by ActiveChooN** [comment]
Got this error, when trying to submit the review
изображение

**Comment by bsekachev** [comment]
@ActiveChooN 
It was merge issue, fixed.

**Comment by nmanovic** [comment]
image

**Comment by bsekachev** [comment]
@nmanovic 

The same issue like above.

**Comment by nmanovic** [comment]
@bsekachev ,

- "Send annotations" isn't very intuitive item of the menu. Probably "Request a review" is better
- In validation mode I see "standard" workspace by default. Probably it is better to have "review" workspace by default.
- In standard workspace if I press "Submit review", nothing happens. Only if I'm in "review" workspace.

Could you please organise a review session this or next week and let's collect some trivial ideas how to polish the initial version and commit. I believe some small adjustments can fix a lot of usability issues. Probably I don't understand how to use it for now.

**Comment by nmanovic** [comment]
What is the reviewer wants to show how to annotate and object. How to do that? It isn't obvious that the comment icon allows to draw something (in our case only bounding box).

**Comment by ActiveChooN** [comment]
@bsekachev, when i trying to change reviewer for a job cvat-core sends empty PATCH request. Also reviewer field is null, but status is "validation".

**Comment by nmanovic** [comment]
Probably submit a review should be. on the review panel as well.

**Comment by bsekachev** [comment]
@nmanovic 

> "Send annotations" isn't very intuitive item of the menu. Probably "Request a review" is better

Not "Send", "Submit". "Submit annotations", "Submit review" looks consistent IMHO.

> In validation mode I see "standard" workspace by default. Probably it is better to have "review" workspace by default.

Review workspace is opened by default for a reviewer, please check if your user was assigned to review

> In standard workspace if I press "Submit review", nothing happens. Only if I'm in "review" workspace.

Okay, it looks like a bug.

**Comment by bsekachev** [comment]
@nmanovic 

> What is the reviewer wants to show how to annotate and object? How to do that? It isn't obvious that the comment icon allows to draw something (in our case only bounding box).

Initially we didn't discuss such the feature. Comment icon is a tool to highlight what region issue related with.

**Comment by bsekachev** [comment]
> @bsekachev, when i'm trying to change reviewer for a job, cvat-core sends empty PATCH request. Also reviewer field is null, but status is "validation".

The same reason, updatedFlag wasn't set when updating reviewer. Fixed.

**Comment by bsekachev** [comment]
> Could you please organise a review session this or next week and let's collect some trivial ideas how to polish the initial version and commit. I believe some small adjustments can fix a lot of usability issues. Probably I don't understand how to use it for now.

Do you mean review of the user interface or the code?

**Comment by nmanovic** [comment]
> @nmanovic
> 
> > Could you please organise a review session this or next week and let's collect some trivial ideas how to polish the initial version and commit. I believe some small adjustments can fix a lot of usability issues. Probably I don't understand how to use it for now.
> 
> Do you mean review of the user interface or the code?

I mean UI/UX. I believe with small changes we can improve it.

**Comment by nmanovic** [comment]
> @nmanovic
> 
> > "Send annotations" isn't very intuitive item of the menu. Probably "Request a review" is better
> 
> Not "Send", "Submit". "Submit annotations", "Submit review" looks consistent IMHO.
> 
> > In validation mode I see "standard" workspace by default. Probably it is better to have "review" workspace by default.
> 
> Review workspace is opened by default for a reviewer, please check if your user was assigned to review
> 
> > In standard workspace if I press "Submit review", nothing happens. Only if I'm in "review" workspace.
> 
> Okay, it looks like a bug.

It can be discussed on the review. When I press "Save" button, I submit annotations. In this case I request a review from somebody. "Request a review" explains what I'm going to do.

I don't think that the view should depend on a user who open the task in "validation" state. It should depend on the state itself. Probably it can be discussed on the review.

**Comment by nmanovic** [comment]
I will summarize my comments/questions for today review in one place:
- How to mark a task as complete without review? In some cases people don't want to have the extra step.
- Rename "Submit annotations" to "Request a review". Need to move the button somewhere. It will be difficult to find.
- It is not clear that button for comments on review panel can draw something. From my perspective the workflow isn't very clear. Let's discuss minimum UI changes to make it more intuitive.
- I will add some useful button on review panel like "hide all objects", "hide all comments" and something like that. Probably it is for second iteration of the feature, submit review, etc.
-  Probably comments should be placed on a free space if it is possible (sometimes they can be on other bounding boxes and it is unclear)
- Probably by right mouse click we have to be able comment something on the image. Now we show an object panel.

**Comment by nmanovic** [comment]
I got the error below. Try to assign an issue on an object and after that remove the object. Go to next frame with an object.

```
TypeError
Cannot read property 'label' of undefined

TypeError: Cannot read property 'label' of undefined
    at Function.mapToProps (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:365:1027292)
    at a (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:365:465545)
    at Function.a.mapToProps (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:365:465682)
    at a (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:365:465545)
    at http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:365:466910
    at http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:365:464319
    at Object.useMemo (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:341:62782)
    at t.useMemo (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:325:5984)
    at d (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:365:464263)
    at qo (http://localhost:8080/cvat-ui.4ab42c810efc6fc7738e.min.js:341:57930)
```

**Comment by bsekachev** [comment]
@nmanovic 

I found another way to reproduce a bit, but anyway should be fixed now.

**Comment by bsekachev** [comment]
@nmanovic @ActiveChooN 
So, the patch looks quite stable. At least I do not see any critical issues any more

**Comment by nmanovic** [comment]
@bsekachev , I will review right now and merge if I don't see any issues.

**Pull request merged**